data/rat.lean

```diff
@@ -931,4 +931,59 @@ theorem nat_ceil_lt_add_one {q : ℚ} (hq : q ≥ 0) : ↑(nat_ceil q) < q + 1 :
 lt_nat_ceil.1 $ by rw [
   show nat_ceil (q+1) = nat_ceil q+1, from nat_ceil_add_nat hq 1]; apply nat.lt_succ_self
 
+@[simp] lemma denom_neg_eq_denom : ∀ q : ℚ, (-q).denom = q.denom 
+| ⟨_, d, _, _⟩ := rfl
+
+@[simp] lemma num_neg_eq_neg_num : ∀ q : ℚ, (-q).num = -(q.num)
+| ⟨n, _, _, _⟩ := rfl 
+
+@[simp] lemma num_zero : rat.num 0 = 0 := rfl 
+
+lemma zero_of_num_zero {q : ℚ} (hq : q.num = 0) : q = 0 :=
+have q = q.num /. q.denom, from num_denom _,
+by simpa [hq]
+
+lemma num_ne_zero_of_ne_zero {q : ℚ} (h : q ≠ 0) : q.num ≠ 0 :=
```

Comment:
I hesitate to bring this up because we've disagreed on this in the past, but you appear to have added a lot of contrapositive theorems like this one. I would prefer not to have duplicate theorems if you can access them with a combination of `.1`, `.2` and `mt` on a base theorem (biconditional if possible). I would demonstrate that whatever use you have for these theorems can be written shorter without them, but I suspect these theorems have received little to no use yet so it would not be a particularly effective demonstration. What are your thoughts on this?

Reply:
Every time I use a biconditional, I have to remember (1) which direction does the name go, (2) do I need .1 or .2, (3) are there arguments that force me to put it in parentheses? It takes me about as long to run through this list as it does to state the contrapositive theorem, and then applying it is mechanical. That said, your style guidelines rule here. If you prefer the concision over my mental energy, that's fine!